jsonComposer.go

```diff
@@ -0,0 +1,218 @@
+			fallthrough
+		case tknArrayStart:
+			depth++
+			composer.Write(tkn)
+			continue
+		case tknObjectEnd:
+			fallthrough
+		case tknArrayEnd:
+			depth--
+			if depth < 0 {
+				return 0, 0, false, fmt.Errorf("invalid JSON object")
+			}
+			composer.Write(tkn)
+			continue
+		case tknEnd:
+			continue
+		default:
+			composer.Write(tkn)
+			continue
+		}
+
+		// strip off the quotes from the string
+		key := tkn[1 : tknLen-1]
+
+		matched := false
+		for _, keyToRemove := range remove {
+			if BytesEqualsString(key, keyToRemove) {
+				matched = true
+
+				// ":"
+				tknType, _, _, err = tokenizer.Step()
+				if err != nil || tknType != tknObjectKeyDelim {
+					return 0, 0, false, fmt.Errorf("error stepping to next token, expecting :, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+				}
+
+				// parse the corresponding value
+				valStart := tokenizer.Position()
+				valEnd := valStart
+				valueDepth := 0
+				done := false
+
+				for !done {
+					tknType, _, _, err = tokenizer.Step()
+					if err != nil {
+						return 0, 0, false, fmt.Errorf("error stepping to next token, expecting JSON value, src=%s, pos=%v, err=%v", src, tokenizer.Position(), err)
+					}
+
+					valEnd = tokenizer.Position()
+
+					switch tknType {
+					case tknObjectStart:
+						fallthrough
+					case tknArrayStart:
+						valueDepth++
+					case tknObjectEnd:
+						fallthrough
+					case tknArrayEnd:
+						valueDepth--
+						if valueDepth == 0 {
+							done = true
+						} else if valueDepth < 0 {
+							return 0, 0, false, fmt.Errorf("invalid JSON object, src=%s, pos=%v", src, tokenizer.Position())
+						}
+					case tknString:
+						fallthrough
+					case tknEscString:
+						fallthrough
+					case tknInteger:
+						fallthrough
+					case tknNumber:
+						fallthrough
+					case tknNull:
+						fallthrough
+					case tknTrue:
+						fallthrough
+					case tknFalse:
+						if valueDepth == 0 {
+							done = true
+						}
+					case tknEnd:
+						return 0, 0, false, fmt.Errorf("unexpected EOF, src=%s, pos=%v", src, tokenizer.Position())
+					default:
+						// tknListDelim, tknObjectKeyDelim, tknUnknown
+					}
+				}
+				removed[keyToRemove] = src[valStart:valEnd]
+				removedLen++
+
+				// can be tknObjectEnd or tknListDelim
+				// if it is tknObjectEnd, we have to step back, remove a tknListDelim and place a tknObjectEnd before commiting
+				// if it is tknListDelim, don't write it
+				tknType, tkn, _, err = tokenizer.Step()
+				if err != nil || (tknType != tknObjectEnd && tknType != tknListDelim) {
+					return 0, 0, false, fmt.Errorf("error stepping to next token, expecting separator or objectEnd, got=%s, src=%s, pos=%v", tkn, src, tokenizer.Position())
+				}
+
+				if tknType == tknObjectEnd {
+					// this was the last item
+					if depth == 1 {
+						removedLastItem = true
+					}
+					depth--
+					if depth < 0 {
+						return 0, 0, false, fmt.Errorf("invalid JSON object")
+					}
+				}
+			}
+		}
+
+		if !matched {
```

Comment:
Logically, does this work if the key to be removed is *inside* an object that is not to be removed? i.e. embedded object. 

Take for example this JSON object:

```
{
   "notKeyToBeRemoved" : {
      "keyToBeRemoved": true
   }
}
```

1. Tokenize first `{`, depth 1
2. Hit `tknString`, check if key to remove
3. Not key to remove... `!matched` is true - "ok to write this item"

The key to be removed is not removed.

Reply:
My intension was to match only depth 1 keys. Hence I have the check:
```
                        // depth 1 strings are JSON object keys, need to check if it is the key to remove
			if depth != 1 {
				composer.Write(tkn)
				continue
			}
```
to skip depth != 1 keys.

Reply:
It is not clear from the function naming that this function only deals with the keys at level 1 depth.

With that said, given this is a library that we're enhancing, I would prefer to have seen the implementation work for all JSON objects. This is because `gojsonsm` has always been working with JSON regardless of depth. I would think this one-off would go against the design principle here for this package.

CC @brett19  for comments.

Reply:
Added support for removing nested keys as well.